superset/connectors/sqla/models.py

```diff
@@ -1124,7 +1125,25 @@ def adhoc_column_to_sqla(
             schema=self.schema,
             template_processor=template_processor,
         )
-        sqla_column = literal_column(expression)
+        col_in_metadata = self.get_column(expression)
+        if col_in_metadata:
+            sqla_column = col_in_metadata.get_sqla_col()
+            is_dttm = col_in_metadata.is_temporal
+        else:
+            sqla_column = literal_column(expression)
+            # probe adhoc column type
+            tbl, _ = self.get_from_clause(template_processor)
+            qry = sa.select([sqla_column]).limit(1).select_from(tbl)
+            sql = self.database.compile_sqla_query(qry)
+            col_desc = get_columns_description(self.database, sql)
+            is_dttm = col_desc[0]["is_dttm"]
+
+        if col.get("is_axis") and col.get("time_grain") and is_dttm:
+            sqla_column = self.db_engine_spec.get_timestamp_expr(
+                sqla_column,
+                None,
+                col.get("time_grain"),
+            )
```

Comment:
I haven't thought quite this far yet, but is there a reason we want to have `is_axis` here in addition to having `time_grain`? I would assume the frontend simply omits `time_grain` if it doesn't want to apply a time grain. Also, mini nit as we're now on 3.8+..
```suggestion
        if col.get("is_axis") and time_grain := col.get("time_grain") and is_dttm:
            sqla_column = self.db_engine_spec.get_timestamp_expr(
                sqla_column,
                None,
                time_grain,
            )
```

Reply:
The `timeGrian` means, this column can use some DB expression (I actually started with a `config` field), but Superset doesn't support other database function mappings now. So I just use the old naming style - 'timeGrain`.

I have changed `is_axis` to the `columnType`. It is easy to extend the type of column using this approach.

Reply:
`:=`  is an awesome new operator!